synapse/storage/databases/main/roommember.py

```diff
@@ -1131,12 +1133,32 @@ async def _get_joined_hosts(
             else:
                 # The cache doesn't match the state group or prev state group,
                 # so we calculate the result from first principles.
+                #
+                # We need to fetch all hosts joined to the room according to `state` by
+                # inspecting all join memberships in `state`. However, if the `state` is
+                # relatively recent then many of its events are likely to be held in
+                # the current state of the room, which is easily available and likely
+                # cached. We therefore compute the set of `state` events not in the
+                # current state and only fetch those.
+                current_state = await self._storage_controllers.state.get_current_state(
+                    room_id
+                )
```

Comment:
Hmm, this will also pull out all the events, which is a bit heavy weight. I think we might want to just do a `SELECT event_id, membership FROM current_state_events WHERE room_id = ?` instead?

If we do use `_storage_controllers.state` we probably want to make sure we don't wait for the full state?

Reply:
Oh I see. I thought you wanted me to use a currently-`@cached` method. But Erik has pointed out that we won't hit this path unless there's been a state change... at which point we won't hit the cache of the current room state.